### ISSUE TITLE
SqlServerDsc: PowerShell DSC resource is not a valid Desired State Configuration resource

### ISSUE THREAD

**Issue by GerbenWelter** [PowerShell/SqlServerDsc#1243]
#### Details of the scenario you tried and the problem that is occurring
We are using the SqlServerDsc code through Puppet and it has been working for some time. Suddenly new machines provisioned with that same code started giving problems. OS and SQL patch levels haven't changed and are the same on working and non-working servers. I have ruled out Puppet because when running the DSC code directly I get the same error. The error is also not very descriptive (to me atleast)

#### Verbose logs showing the problem
The C:/ProgramData/PuppetLabs/puppet/cache/lib/puppet_x/dsc_resources/SqlServerDsc/SqlServerDsc.psd1 PowerShell DSC resource is not a valid Desired State Configuration 
resource.
    + CategoryInfo          : InvalidOperation: (root/Microsoft/...gurationManager:String) [], CimException
    + FullyQualifiedErrorId : ProviderNotFound
    + PSComputerName        : localhost

#### Suggested solution to the issue
n/a

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
This is the code that Puppet generates and can be run  standalone in Powershell to trigger the error.

```powershell
function new-pscredential
{
  [CmdletBinding()]
  param (
    [parameter(Mandatory=$true,
      ValueFromPipelineByPropertyName=$true)]
    [string]
    $user,

    [parameter(Mandatory=$true,
      ValueFromPipelineByPropertyName=$true)]
    [string]
    $password
  )

  $secpasswd   = ConvertTo-SecureString $password -AsPlainText -Force
  $credentials = New-Object System.Management.Automation.PSCredential ($user, $secpasswd)
  return $credentials
}

$response = @{
  indesiredstate = $false
  rebootrequired = $false
  errormessage   = ''
}

$invokeParams = @{
  Name          = 'SqlServerConfiguration'
  Method        = 'test'
  Property      = @{
    psdscrunascredential = [PSCustomObject]@{'user' = 'DOMAIN\SA_Account'; 'password' = 'Secret'} | new-pscredential
    servername = 'ServerName'
    instancename = 'MSSQLSERVER'
    optionname = 'xp_cmdshell'
    optionvalue = 0
  }
  ModuleName = @{
    ModuleName      = "C:/ProgramData/PuppetLabs/puppet/cache/lib/puppet_x/dsc_resources/SqlServerDsc/SqlServerDsc.psd1"
    RequiredVersion = "11.1.0.0"
  }
}

Invoke-DscResource @invokeParams
```

Output on working server:

```
InDesiredState 
-------------- 
True           
```

Output on failing server:

```
The C:/ProgramData/PuppetLabs/puppet/cache/lib/puppet_x/dsc_resources/SqlServerDsc/SqlServerDsc.psd1 PowerShell DSC resource is not a valid Desired State Configuration 
resource.
    + CategoryInfo          : InvalidOperation: (root/Microsoft/...gurationManager:String) [], CimException
    + FullyQualifiedErrorId : ProviderNotFound
    + PSComputerName        : localhost
```

Why does it fail on the psd1 and not on some psm1? What does the module need as a provider?

#### SQL Server edition and version the target node is running

Version through Powershell:

Major         : 13
Minor         : 0
Build         : 5081
Revision      : -1
MajorRevision : -1
MinorRevision : -1

#### SQL Server PowerShell modules present on the target node

Name         Version Path                                                                                                   
----         ------- ----                                                                                                   
SqlServer    20.0    C:\Program Files\WindowsPowerShell\Modules\SqlServer\SqlServer.psd1                                    
SQLASCMDLETS 2.0     C:\Program Files (x86)\Microsoft SQL Server\130\Tools\PowerShell\Modules\SQLASCMDLETS\SQLASCMDLETS.psd1
SQLPS        1.0     C:\Program Files (x86)\Microsoft SQL Server\130\Tools\PowerShell\Modules\SQLPS\SQLPS.psd1

#### The operating system the target node is running

OsName               : Microsoft Windows Server 2016 Standard
OsOperatingSystemSKU : StandardServerEdition
OsArchitecture       : 64-bit
WindowsBuildLabEx    : 14393.2485.amd64fre.rs1_release.180827-1809
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### Version and build of PowerShell the target node is running

Name                           Value                                                                                                                                     
----                           -----                                                                                                                                     
PSVersion                      5.1.14393.2485                                                                                                                            
PSEdition                      Desktop                                                                                                                                   
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                   
BuildVersion                   10.0.14393.2485                                                                                                                           
CLRVersion                     4.0.30319.42000                                                                                                                           
WSManStackVersion              3.0                                                                                                                                       
PSRemotingProtocolVersion      2.3                                                                                                                                       
SerializationVersion           1.1.0.1  

#### Version of the DSC module that was used ('dev' if using current dev branch)

Version 11.1.0.0 (coming from the Puppet DSC module version 1.6.0)

**Comment by johlju** [comment]
Maybe there is something in the .psd1 file making unable to parse. 🤔 Can you please try to reproduce the problem  with the latest module in PowerShell Gallery?

**Comment by GerbenWelter** [comment]
As said earlier, this exact code is ditributed to other machines where it runs succesfully. Something seems to have changed in the way we deploy those newer systems causing the this code to fail. But why on the psd1 file with the ProviderNotFound error? What does DSC see as a provider in this code?

I'll have a go with the latest version to see if that changes anything.

**Comment by johlju** [comment]
There are no provider needed to import the module. Only provider that is imported is when the resources import either the SQLPS or SqlServer module (SQLSERVER: provider part of those modules).

Do you get the same problem when importing the module in a regular PowerShell console?

**Comment by GerbenWelter** [comment]
I just imported the module directly in the powershell console and that works as expected. The problem is that we have a set of working servers and a set of non-working servers. We just can't pinpoint the difference causing the problem.

I have tried following the code regarding SQLPS and SqlServer but I haven't found what code triggers the error. Debugging DSC code is harder than it should be. The problem is compounded by using DSC through Puppet making it harder to replicate.

I'm closing this issue because a) I was hoping to get some information about the provider but as you have stated there is no provider needed and b) this problem seems very specific to our Puppet environment. We just need to get to the bottom of this. Anyway, thanks for the help.
